### PR TITLE
Allow attribute values for access strategy to support regex

### LINKED ISSUE #1414: Wildcard/Regex support in access control
For example, the documentation shows an example access control situation that lists valid values of the required attribute:

    "cn" : [ "java.util.HashSet", [ "admin, Admin, TheAdmin" ] ]

Could include support for a wildcard feature or full-blown regex such as:

    "cn" : [ "java.util.HashSet", [ ".*[Aa]dmin.*" ] ]

Not that you'd trust your admin access control to such a filter but as an example of possible capability.

### PR DESCRIPTION
**Pull request by mmoayyed**
Closes #1414

### CODE REVIEW

File: cas-server-core-util/src/main/java/org/jasig/cas/util/RegexUtils.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+/**
+ * Utility class to assist with resource operations.
```
Comment (dima767) [inline cas-server-core-util/src/main/java/org/jasig/cas/util/RegexUtils.java]: Probably copy/paste javadoc comment from ResourceUtils ;-)

Comment (mmoayyed) [inline cas-server-core-util/src/main/java/org/jasig/cas/util/RegexUtils.java]: Yup. You caught me :) I will clean this up.

File: cas-server-core-util/src/main/java/org/jasig/cas/util/RegexUtils.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+     * Concatenate all elements in the given collection to form a regex pattern.
+     *
+     * @param requiredValues  the required values
+     * @param caseInsensitive the case insensitive
+     * @return the pattern
+     */
+    public static Pattern concatenate(final Collection<String> requiredValues, final boolean caseInsensitive) {
+        final StringBuilder builder = new StringBuilder(requiredValues.size());
+        for (final String requiredValue : requiredValues) {
+            builder.append('(').append(requiredValue).append(")|");
```
Comment (frett) [inline cas-server-core-util/src/main/java/org/jasig/cas/util/RegexUtils.java]: do we need to do any type of escaping on `requiredValue` so it's not processed as regex special characters?

Comment (frett) [inline cas-server-core-util/src/main/java/org/jasig/cas/util/RegexUtils.java]: actually, strike that. I see that requirement was part of the original issue

Comment (leleuj) [inline cas-server-core-util/src/main/java/org/jasig/cas/util/RegexUtils.java]: I'm in line with Daniel: isn't escaping required?

Comment (mmoayyed) [inline cas-server-core-util/src/main/java/org/jasig/cas/util/RegexUtils.java]: Escaping is required, not by us. By the deployer. Either the pattern is a valid regex or it's not. The deployer has to make sure it's valid. If it isn't, we don't treat it as such and we don't try to guess the deployer's intention.

**Comment by leleuj** [comment]
+1

File: cas-server-core-util/src/main/java/org/jasig/cas/util/RegexUtils.java (hunk 2 above)
Comment (mmoayyed) [inline cas-server-core-util/src/main/java/org/jasig/cas/util/RegexUtils.java]: To further explain this, let say we have the following rule:

```
"cn" : [ "java.util.HashSet", [ ".*[Aa]dmin.*" ] ]
```

The deployer then decides to add an extra "." catching `Admin.Misagh` for example. So the pattern becomes:

```
"cn" : [ "java.util.HashSet", [ ".*[Aa]dmin\..*" ] ]
```

So you see, I as the deployer have the responsibility of escaping. Technically, if I don't escape it, it's still a valid regex. But nonetheless, that's my job. Not CAS.

Comment (mmoayyed) [inline cas-server-core-util/src/main/java/org/jasig/cas/util/RegexUtils.java]: If you did however put in:

```
"cn" : [ "java.util.HashSet", [ ".*[Aa]dmin.****" ] ]
```

Then, that is not valid regex as the pattern won't compile. Thus, we treat it as a verbatim attribute value

**Pull request merged**